### ISSUE TITLE
parametric/symbolic evaluation in quantum_info

### ISSUE THREAD

**Issue by ajavadia** [Qiskit/qiskit-terra#4751]
A way to obtain Operators and evolve Statevectors in the `qiskit.quantum_info` module where some parameters are unknown. Example:

```py
from qiskit.circuit import QuantumCircuit, ParameterVector
circ = QuantumCircuit(2)
theta = ParameterVector(name='θ', length=17)
circ.u3(theta[0], theta[1], theta[2], 0)
circ.u3(theta[3], theta[4], theta[5], 1)
circ.cx(0, 1)
circ.h(0)
circ.rz(theta[6], 1)
circ.h(1)
circ.rz(theta[7], 1)
circ.cx(0, 1)
circ.rz(theta[8], 0)
circ.h(0)
circ.rz(theta[9], 1)
circ.h(1)
circ.rz(theta[10], 1)
circ.cx(0, 1)
circ.u3(theta[11], theta[12], theta[13], 0)
circ.u3(theta[14], theta[15], theta[16], 1)
circ.draw('mpl', fold=500, scale=2)
```

![image](https://user-images.githubusercontent.com/8622381/87854973-67a1a680-c8e3-11ea-9048-50f4f845e977.png)

This would be a 4x4 matrix whose parameters are mathematical expressions of `θ`.

**Comment by a-kapila** [comment]
Hi, I'd like to work on this if possible - would it be OK to use eg SymPy to perform symbolic calculation?

**Comment by ajavadia** [comment]
Great. Yeah I think Sympy is good but we should make it work with Qiskit’s parameterized circuits. So one would be able to do normal calculations like Operator(circuit), Statevector.evolve(circuit) but with parameters

**Comment by a-kapila** [comment]
Cheers @ajavadia I'll update here once I've got something working

**Comment by georgios-ts** [comment]
Hi @ajavadia ,@a-kapila  are you still working on it? 
I took a look at this issue and i realized that now the code crashes after calling `to_matrix()` method of standard gates since the method expects the `Parameter`s of the Gate to be binded. You can make this work but you will not be able to return a numpy array with complex entries. (the entries will be of `dtype=object`). The issue is that the `Operator` class strongly assumes that the `_data` field will be a numpy array with complex entries. So, i think it's better to make a new class rather than to mess a lot with the existing one. Any thoughts?

**Comment by a-kapila** [comment]
Hi @georgios-ts sorry, not working on it any more - found a workaround for what I needed so feel free to go for it yourself.

**Comment by ajavadia** [comment]
@georgios-ts @a-kapila sorry for the late response. this is still a good issue if you are interested to work on it. I was hoping we could make the Operator more flexible to consume unbound parameters (i.e. sympy values). But I haven't dug into much. @chriseclectic thoughts?

**Comment by georgios-ts** [comment]
@ajavadia Yes, still interested. Though i think it's possible only after #4854 and #5192 merge, since they are necessary to express, for example, the entries of a rotation matrix with unbound parameters.

**Comment by chriseclectic** [comment]
As @gerogios-ts says you can't just swap the numpy array for sympy as many things will break. I think it would be better to make a new parameterized operator class that matches the Operator API but can also be initialized from sympy array or parameterized quantum circuit and has extra methods for setting parameter values and converting to numeric operator.

**Comment by chriseclectic** [comment]
One issue I would like to point out is whether it is if it is even worth doing something like this symbolically in the first place. One you try and convert a parameterized circuit to a symbolic matrix the expressions are going to blow up very fast.

To take the circuit from the example and evaluating it symbolically in Mathematica, the U[0, 0] entry is:
```
1/4 E^(-(1/2)
    I (\[Theta][6] + \[Theta][7] + \[Theta][8] + \[Theta][
     9] + \[Theta][10])) (Cos[\[Theta][0]/
     2] (Cos[\[Theta][3]/
        2] (Cos[\[Theta][11]/
           2] ((1 + E^(I (\[Theta][7] + \[Theta][8])) + E^(
               I (\[Theta][7] + \[Theta][9])) + E^(
               I (\[Theta][8] + \[Theta][9]))) Cos[\[Theta][14]/2] - 
            E^(I (\[Theta][10] + \[Theta][16])) (1 + E^(
               I (\[Theta][7] + \[Theta][8])) - E^(
               I (\[Theta][7] + \[Theta][9])) - E^(
               I (\[Theta][8] + \[Theta][9]))) Sin[\[Theta][14]/2]) + 
         E^(I \[Theta][13])
           Sin[\[Theta][11]/
           2] (E^(I \[Theta][10]) (-1 + E^(
               I (\[Theta][7] + \[Theta][8])) + E^(
               I (\[Theta][7] + \[Theta][9])) - E^(
               I (\[Theta][8] + \[Theta][9]))) Cos[\[Theta][14]/2] - 
            E^(I \[Theta][16]) (-1 + E^(
               I (\[Theta][7] + \[Theta][8])) - E^(
               I (\[Theta][7] + \[Theta][9])) + E^(
               I (\[Theta][8] + \[Theta][9]))) Sin[\[Theta][14]/2])) -
       E^(I (\[Theta][4] + \[Theta][6]))
        Sin[\[Theta][3]/
        2] (E^(I \[Theta][13])
           Sin[\[Theta][11]/
           2] (E^(I \[Theta][10]) (1 + E^(
               I (\[Theta][7] + \[Theta][8])) + E^(
               I (\[Theta][7] + \[Theta][9])) + E^(
               I (\[Theta][8] + \[Theta][9]))) Cos[\[Theta][14]/2] - 
            E^(I \[Theta][16]) (1 + E^(
               I (\[Theta][7] + \[Theta][8])) - E^(
               I (\[Theta][7] + \[Theta][9])) - E^(
               I (\[Theta][8] + \[Theta][9]))) Sin[\[Theta][14]/2]) + 
         Cos[\[Theta][11]/
           2] ((-1 + E^(I (\[Theta][7] + \[Theta][8])) + E^(
               I (\[Theta][7] + \[Theta][9])) - E^(
               I (\[Theta][8] + \[Theta][9]))) Cos[\[Theta][14]/2] - 
            E^(I (\[Theta][10] + \[Theta][16])) (-1 + E^(
               I (\[Theta][7] + \[Theta][8])) - E^(
               I (\[Theta][7] + \[Theta][9])) + E^(
               I (\[Theta][8] + \[Theta][9]))) Sin[\[Theta][14]/
              2]))) + 
   E^(I \[Theta][1])
     Sin[\[Theta][0]/
     2] (E^(I \[Theta][4])
        Sin[\[Theta][3]/
        2] (Cos[\[Theta][11]/
           2] (-(-1 + E^(I (\[Theta][7] + \[Theta][8])) - E^(
                I (\[Theta][7] + \[Theta][9])) + E^(
                I (\[Theta][8] + \[Theta][9]))) Cos[\[Theta][14]/2] + 
            E^(I (\[Theta][10] + \[Theta][16])) (-1 + E^(
               I (\[Theta][7] + \[Theta][8])) + E^(
               I (\[Theta][7] + \[Theta][9])) - E^(
               I (\[Theta][8] + \[Theta][9]))) Sin[\[Theta][14]/2]) - 
         E^(I \[Theta][13])
           Sin[\[Theta][11]/
           2] (E^(I \[Theta][10]) (1 + E^(
               I (\[Theta][7] + \[Theta][8])) - E^(
               I (\[Theta][7] + \[Theta][9])) - E^(
               I (\[Theta][8] + \[Theta][9]))) Cos[\[Theta][14]/2] - 
            E^(I \[Theta][16]) (1 + E^(
               I (\[Theta][7] + \[Theta][8])) + E^(
               I (\[Theta][7] + \[Theta][9])) + E^(
               I (\[Theta][8] + \[Theta][9]))) Sin[\[Theta][14]/2])) +
       E^(I \[Theta][6])
        Cos[\[Theta][3]/
        2] (E^(I \[Theta][13])
           Sin[\[Theta][11]/
           2] (E^(I \[Theta][10]) (-1 + E^(
               I (\[Theta][7] + \[Theta][8])) - E^(
               I (\[Theta][7] + \[Theta][9])) + E^(
               I (\[Theta][8] + \[Theta][9]))) Cos[\[Theta][14]/2] - 
            E^(I \[Theta][16]) (-1 + E^(
               I (\[Theta][7] + \[Theta][8])) + E^(
               I (\[Theta][7] + \[Theta][9])) - E^(
               I (\[Theta][8] + \[Theta][9]))) Sin[\[Theta][14]/2]) - 
         Cos[\[Theta][11]/
           2] ((-1 - E^(I (\[Theta][7] + \[Theta][8])) + E^(
               I (\[Theta][7] + \[Theta][9])) + E^(
               I (\[Theta][8] + \[Theta][9]))) Cos[\[Theta][14]/2] + 
            E^(I (\[Theta][10] + \[Theta][16])) (1 + E^(
               I (\[Theta][7] + \[Theta][8])) + E^(
               I (\[Theta][7] + \[Theta][9])) + E^(
               I (\[Theta][8] + \[Theta][9]))) Sin[\[Theta][14]/2]))))
```

**Comment by georgios-ts** [comment]
@chriseclectic Good point. I think it's only valuable in circuits with a small number of qubits and parameters. For example, it's annoying when i have a 2q-circuit with a single unbound parameter and cannot inspect the unitary.

**Comment by mrvee-qC** [comment]
# Update for Santa - QGoGP

:white_check_mark: Issue still valid! Statevector class and operators are not flexible enough to evolve parameterised operations. No PRs linked to this issue that have solved it completely as per OP's description. Issue is open to solve!

Will update comment/add more comments if cause is found/fixed!

> Python version 3.9.7
> qiskit-terra version: 0.19.1

**Comment by Cryoris** [comment]
If we implement this we might add a very big warning about the performance of this (and the nasty results). If a user starts looking at the unitary of some variational circuit for instance, this feature might create more frustration than it is useful 😅

**Comment by evmckinney9** [comment]
Hmm interesting problems. Here is what I'm thinking about on how to do this programmatically. 

```
def get_symb_express(base):
    class SymbolicGate(base):
        def __init__(self):
            pass
        
        def eval(self, symb):
            self.params = [symb]
            #import sympy as math
            return self.__array__()

        @property
        def params(self):
            """return instruction params."""
            return self._params

        #remove check to validate_params()
        @params.setter
        def params(self, parameters):
            self._params = parameters

    theta = sympy.Symbol("theta")
    return SymbolicGate().eval(theta)

get_symb_express(RXGate)
```

or something more direct

```
class foo():
    def __init__(self):
        self.params = [sympy.Symbol("theta")]
foo.__array__ = RXGate.__array__.__get__(foo())
foo.__array__()
```

Even if you are able to figure out the inheritance problem so that the parent classes of Gate decide to use the sympy library instead of math library for `math.cos` and `math.sin` functions (by doing something equivalent to `math=sympy` somewhere) , you still get problems for gates like RZGate, that cast params as float. I tried making a wrapper class on `sympy.Symbol` which overrides the float method to return itself, but python throws a type error and doesn't allow this.

Now a really ugly solution I'm thinking about is to use `inspect.getsource()` to grab the __array__ definition from the parent class, but rewrite it so it removes calls to `float` and replaces `math.` with `sympy.`. I'm surprised there is no easier way to do this, but all the type-checking is making it tough. I'd very much like to be able to multiply operator matrices together without needing to instantiate gates with parameter values.

**Comment by evmckinney9** [comment]
Ugly code but proof of concept:
```
from inspect import getsource
import re

class SymbolicGate():
    def __init__(self):
        pass
        
    def eval(self, base, symb):
        self.params = [symb]
        function_string =  getsource(base.__array__)
        function_string = function_string.replace("return", "self._temp=")
        function_string = function_string.replace("math", "sympy")
        function_string = function_string.replace(", dtype=dtype", "")
        function_string = function_string.replace("np.exp", "sympy.exp")
        function_formatted = ""
        for line in function_string.splitlines():
            if "def" in line:
                continue
            else:
                if "float" in line:
                    line = line.replace("float(", "")
                    line = line[:-1]
                #shift over from class indentation
                function_formatted += line[8:]
                function_formatted += "\n"
        
        #function_formatted = "import sympy\n" + function_formatted
        exec(function_formatted, {"sympy":sympy, "numpy":np, "self":self})
        return self._temp


theta = sympy.Symbol("theta")
print(SymbolicGate().eval(RXGate, theta))
print(SymbolicGate().eval(RZGate, theta))
```
Outputs:

[[cos(theta/2) -1.0*I*sin(theta/2)]
 [-1.0*I*sin(theta/2) cos(theta/2)]]

[[exp(-0.5*I*theta) 0]
 [0 exp(0.5*I*theta)]]

**Comment by Cryoris** [comment]
Another option might be that we do not allow the symbolic evaluation of circuits, but instead allow casting parameterized gates to a matrix. Then, the user would have to use SymPy directly to construct the parameterized unitary.

**Comment by SimoneGasperini** [comment]
What about defining a specific "sub-library" for the symbolic representation of each quantum gate by using `sympy`? Do you think this would be a viable solution? What I mean is something like the following (limiting to the implementation of some basic single-qubit gates only):

```
from sympy import *
from sympy.matrices import Matrix

class SympyUGate:
    def __init__(self, t, p, l):
        self.t = t
        self.p = p
        self.l = l

    def eval(self):
        return Matrix([[cos(self.t/2), -exp(I*self.l) * sin(self.t/2)],
                       [exp(I*self.p) * sin(self.t/2), exp(I*(self.p+self.l)) * cos(self.t/2)]])

def id(): return SympyUGate(t=0, p=0, l=0).eval()
def x(): return SympyUGate(t=pi, p=0, l=pi).eval()
def y(): return SympyUGate(t=pi, p=pi/2, l=pi/2).eval()
def z(): return SympyUGate(t=0, p=0, l=pi).eval()
def h(): return SympyUGate(t=pi/2, p=0, l=pi).eval()
def rx(theta): return SympyUGate(t=theta, p=-pi/2, l=pi/2).eval()
def ry(theta): return SympyUGate(t=theta, p=0, l=0).eval()
def rz(phi): return SympyUGate(t=0, p=0, l=phi).eval()

symbolic_gates = {'id': id, 'x': x, 'y': y, 'z': z,
                  'h': h, 'rx': rx, 'ry': ry, 'rz': rz}
```

Then it would be possible to get the gate symbolic representation by implementing the method `to_sympy()` to the class [`qiskit.circuit.Gate`](https://qiskit.org/documentation/stubs/qiskit.circuit.Gate.html):

```
from qiskit.circuit.library import *
from qiskit.circuit import Parameter

# add this method to the qiskit.circuit.Gate class
def to_sympy(self):
    return symbolic_gates[self.name](*(par.sympify() for par in self.params))

# example for RX rotation gate
theta = Parameter('θ')
rx = RXGate(theta)
rx.to_sympy()
```
$$
\begin{bmatrix}
\cos\left(\frac{\theta}{2}\right) & -i \sin\left(\frac{\theta}{2}\right)\\ 
-i \sin\left(\frac{\theta}{2}\right) & \cos\left(\frac{\theta}{2}\right)
\end{bmatrix}
$$

**Comment by SimoneGasperini** [comment]
Once this is done, in principle it would be possible to get the symbolic unitary matrix (or the final full statevector) representation out of any given parameterized quantum circuit.
A very raw but working implementation of this transformation (considering again single-qubit gates only) could be:
```
from qiskit.circuit.library import IGate
from sympy import KroneckerProduct

def get_sympy_operator(qc):
    num_layers = qc.depth()
    num_qubits = qc.num_qubits
    qc_layers = [[IGate() for q in range(num_qubits)]
                 for l in range(num_layers)]
    layer_index = [0] * num_qubits
    for gate in qc.data:
        i = gate.qubits[0]._index
        qc_layers[layer_index[i]][i] = gate.operation
        layer_index[i] += 1
    qc_layers = [[qc_layers[l][-q].to_sympy() for q in range(1, num_qubits+1)]
                 for l in range(num_layers)]
    sympy_operator = KroneckerProduct(*qc_layers[0])
    for j in range(1, num_layers):
        sympy_operator = KroneckerProduct(*qc_layers[j]) * sympy_operator
    return sympy_operator.doit()
```

Proof-of-concept example:
```
from qiskit.circuit import QuantumCircuit, Parameter

theta = Parameter('θ')
qc = QuantumCircuit(2)
qc.x(0)
qc.h(1)
qc.rx(theta, 0)

qc.draw('mpl')
```
![output](https://user-images.githubusercontent.com/71086758/209119341-7f055ec2-d87f-4312-b32a-899bcea1fb63.png)

`get_sympy_operator(qc)`

![img](https://user-images.githubusercontent.com/71086758/209119167-2812dae2-f28e-4076-ac55-0cd1447259bb.gif)

**Comment by SimoneGasperini** [comment]
@ajavadia Any comment or suggestion about this? If this approach makes some sense, I would be glad to start working on this

**Comment by SimoneGasperini** [comment]
@ajavadia @jakelishman: any comment or suggestion about this? If this approach makes some sense, I would be glad to start working on this

**Comment by jakelishman** [comment]
I'm very hesitant to add any further symbolic capabilities to Terra - even the `quantum_info` base probably wouldn't have hurt if it were managed as a separate package that's a dependency of Terra rather than inside it.  Terra now is most properly the owner of various API interfaces, and of circuit descriptions, manipulations and compilation, and `quantum_info` is kind of a part we'd rather somebody else could manage.

The way you're describing here sounds like it would be quite separate even to `quantum_info`.  That does sound like the most sensible way to implement something like this.  That also means, though, that it doesn't necessarily need to be in Terra - you could work on packaging it up yourself using Terra as a dependency, and consider submitting it to [the Qiskit Ecosystem](https://qiskit.org/ecosystem/).  If it does become something that we think is good to merge into Terra, it's easier to go from a separate package into Terra than it is to deal with code that's merged to Terra when we don't have the manpower to maintain it.

**Comment by SimoneGasperini** [comment]
Thank you for the suggestion... I started to work on that as a separate package: [SimoneGasperini/qiskit-symbolic](https://github.com/SimoneGasperini/qiskit-symbolic) including `qiskit-terra` as a dependency. If anyone wants to help, please feel free to leave some suggestions here or open a new issue/pull request on the repo itself.

**Comment by jakelishman** [comment]
Thanks Simone - if you're interested in submitting it to the ecosystem I linked above, please feel free to do so, and one of the team there will help you out with the process.

I'll close this issue for now, because we don't really have a clear vision on whether the team as a whole wants to accept / maintain a PR to this effect in Terra.  We may re-open it in the future, but for now I think it's misleading people into thinking we're more willing to accept this than we actually are.
